### ISSUE TITLE
KeyError: b'error'  in code.intern_identifier(self.attribute) with in Cython-3.0.8-nogil39-nogil

### ISSUE THREAD

**Issue by ogrisel** [colesbury/nogil#139]
The nightly scikit-learn build on Python 3.9 nogil recently started to fail at build time (when running Cython) with the following exception:

```python
Exception ignored in tp_clear of: <class 'cell'>
Traceback (most recent call last):
  File "/io/testvenv/lib/python3.9/site-packages/Cython/Compiler/ExprNodes.py", line 7750, in generate_result_code
    code.intern_identifier(self.attribute),
KeyError: b'error'
performance hint: sklearn/tree/_criterion.pyx:1707:41: Exception check after calling '__pyx_fuse_1xlogy' will always require the GIL to be acquired. Declare '__pyx_fuse_1xlogy' as 'noexcept' if you control the definition and you're sure you don't want the function to raise exceptions.
```

Observed in:

- https://dev.azure.com/scikit-learn/scikit-learn/_build/results?buildId=65040&view=logs&j=67fbb25f-e417-50be-be55-3b1e9637fce5&t=b34c796d-89c1-514c-90f8-edfff5e3a76a

Side questions: how long do you plan on "supporting" the 3.9 nogil CPython stack? Do you encourage projects that want to test against `nogil`-CPython to updated their CI setup to use some recent nogil build of the CPython 3.13 dev branch? Or is it still too early to do so?

/cc @lesteve

**Comment by colesbury** [comment]
Do you know if something changed recently? New version of Cython?

I think it's still too early to test with the dev version of CPython 3.13, but we are getting close.

**Comment by lesteve** [comment]
Not too sure why this happened but it looks like the scikit-learn nogil CI is passing in the last few days, so maybe this can be closed?

See for example this [build log](https://dev.azure.com/scikit-learn/scikit-learn/_build/results?buildId=65090&view=logs&j=67fbb25f-e417-50be-be55-3b1e9637fce5&t=b34c796d-89c1-514c-90f8-edfff5e3a76a)

**Comment by ogrisel** [comment]
> New version of Cython?

There is indeed a new version of Cython (3.0.9) that was released quite recently and that we would like to use soon. However the error I reported was a build locked with the 3.0.8 version so this is unrelated.

> Not too sure why this happened but it looks like the scikit-learn nogil CI is passing in the last few days, so maybe this can be closed?

Indeed. This is very weird to have a non-deterministic failure at the Cythonization step.

Let's close. I'll reopen another issue later (maybe on the `nogil-wheels` repo instead) if the problem shows up again.
